### PR TITLE
Fixes #7 - inbound environment variables

### LINKED ISSUE #7: retain backward compatibility for environment variables
The legacy singularity supports environment variables named "SINGULARITY_*" and "SINGULARITYENV_*"

Apptainer supports environment variables named "APPTAINER_*" and "APPTAINERENV_*"

Proposed:

1. When the inherited environment has APPTAINER variable of same name suffix as a set SINGULARITY environment variable, the APPTAINER value shall have precedence. 

2. When a SINGULARITY environment variable is set, and the APPTAINER environment with the same suffix is not set, copy the value of the SINGULARITY environment variable to a newly created APPTAINER variable with same name suffix.

3. When creating the environment for a child process, enumerate all APPTAINER variable values from (1, 2) above, and create a new set of SINGULARITY environment variables with the same value and name suffix as the APPTAINER variable.

### LINKED ISSUE #90: retain backward compatibility for outbound/published environment variables for child/execed processes
@DrDaveD for the inbound environment variables, they go through a single function for the most part (and the remainder should be re-routed via that function)

Also, this issue should be broken into two separate issues.
1. inbound environment variables
2. published environment variables for child/exec'd processes

_Originally posted by @PaulCharlton in https://github.com/apptainer/apptainer/issues/7#issuecomment-998901310_

### PR DESCRIPTION
**Pull request by PaulCharlton**
Fixes #7 - inbound environment variables

also Fixes #90 -- outbound environment variables

### CODE REVIEW

File: internal/pkg/runtime/engine/config/oci/generate/generate.go [hunk 1]
```diff
@@ -110,6 +110,13 @@ func (g *Generator) AddProcessEnv(env, value string) {
 	g.Config.Process.Env = append(g.Config.Process.Env, kenv+value)
 }
 
+// SetProcessEnvWithPrefixes adds or replaces a container process environment variable for each prefix provided
```
Comment (DrDaveD) [inline internal/pkg/runtime/engine/config/oci/generate/generate.go:113]: How about adding one more function named something like `SetProcessPrefixedEnv` which calls `SetProcessEnvWithPrefixes(env.ApptainerPrefixes, ...` to simply those repeated calls.

**Comment by DrDaveD** [comment]
The debian ci failure should be fixed by #110

File: internal/pkg/runtime/engine/config/oci/generate/generate.go (hunk 1 above)
Comment (PaulCharlton) [reply internal/pkg/runtime/engine/config/oci/generate/generate.go]: @DrDaveD time to leave `good enough` alone ... introducing the change you suggest creates a circular reference between modules which takes even more refactoring to resolve

File: internal/pkg/util/env/create.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+						envKeys["SING_USER_DEFINED_PATH"] = e[1]
+						envKeys["APPTAINER_USER_DEFINED_PATH"] = e[1]
+					default:
+						if key != "" {
+							if permitted, ok := alwaysOmitKeys[key]; ok && !permitted {
+								sylog.Warningf("Overriding %s environment variable with %s is not permitted", key, e[0])
+								continue
+							}
+							if value, ok := envKeys[key]; ok {
+								sylog.Verbosef("Skipping environment variable [%s=%s], %s is already overridden with [%s]", e[0], e[1], key, value)
```
Comment (ikaneshiro) [inline internal/pkg/util/env/create.go:72]: Based on the discussion in #7, it looks like this should be `Warningf` so users get output at the default log level

**Review by ikaneshiro — COMMENTED** [review]
Looks good to me overall, one nit

File: internal/pkg/util/env/create.go (hunk 2 above)
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go]: @ikaneshiro addressed with a common function for all environment variable overrides

**Review by cclerget — CHANGES_REQUESTED** [review]

File: internal/pkg/util/env/create.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+			if strings.HasPrefix(env, prefix) {
+				e := strings.SplitN(env, "=", 2)
+				if len(e) != 2 {
+					sylog.Verbosef("Can't process override environment variable %s", env)
+				} else {
+					key := e[0][len(prefix):]
+					if key != "" {
+						switch key {
+						case "PREPEND_PATH":
+							setKeyIfNotAlreadyOverriden(g, envKeys, e[0], "SING_USER_DEFINED_PREPEND_PATH", e[1])
```
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:71]: @cclerget this is not my understanding.   my understanding is that these scripts are bound in various containers and we need to provide the old name for old containers and can have the new containers use the new name at build time and run time.

File: internal/pkg/build/util.go [hunk 4]
```diff
@@ -148,13 +148,17 @@ func currentEnvNoApptainer(permitted []string) []string {
...
-			envs = append(envs, e)
-		} else {
-			envKey := strings.SplitN(e, "=", 2)
-			if slice.ContainsString(permitted, strings.TrimPrefix(envKey[0], env.ApptainerPrefix)) {
-				sylog.Debugf("Passing through env var %s to apptainer", e)
+		for _, prefix := range env.ApptainerPrefixes {
+			if !strings.HasPrefix(e, prefix) {
 				envs = append(envs, e)
+				break
+			} else {
```
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/build/util.go:155]: revised

File: internal/pkg/util/env/create.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		for _, prefix := range ApptainerEnvPrefixes {
+			if strings.HasPrefix(env, prefix) {
+				// already processed overrides
+				continue EnvKeys
+			}
+		}
+
+		// non prefixed environment variables
+		if value, ok := envKeys[e[0]]; ok {
+			sylog.Verbosef("Skipping [%[1]s=%[2]s environment variable, overridden by [%[1]s=%[3]s]", e[0], e[1], value)
```
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: @cclerget this was quite intentional and I do not see the confusion per se.   We are not keeping enough state around to know which prefix was provided to override "FOO=2" was it APPTAINERENV_FOO or SINGULARITYENV_FOO ?    saving additional state and related metadata did not seem worthy of the extra complexity to generate the informational message.

File: internal/pkg/util/env/create.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	sylog.Verbosef("Setting PATH=%s", DefaultPath)
+	g.SetProcessEnv("HOME", homeDest)
+	g.SetProcessEnv("PATH", DefaultPath)
+
+	return envKeys
+}
+
+// testAddToHostEnv processes given key and returns if the environment
+// variable should be added to the container or not.
+func testAddToHostEnv(key string, cleanEnv bool) bool {
```
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:150]: @cclerget I like the ruby syntax for this one ... "func AddToHostEnv?(...)"

revised

File: internal/pkg/util/env/create.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cclerget) [reply internal/pkg/util/env/create.go:71]: `RuntimeVars` script is a virtual one used by the shell interpreter which means that the one in the image is just ignored by the runtime to use this one, so it's safe to rename there without worrying about build. The script used for build of images is located [here](https://github.com/apptainer/apptainer/blob/042b6df18068148365aed054b000f80edd3234f7/internal/pkg/build/sources/base_environment.go#L259)

File: internal/pkg/util/env/create.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cclerget) [reply internal/pkg/util/env/create.go:132]: Ok so something like `overridden by the prefixed environment variable KEY=VAL` ?

File: internal/pkg/util/env/create.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:71]: @cclerget ` the one in the image is just ignored by the runtime`.   If this is true, let's file that as a separate issue, and leave the current "inefficiency" of duplicate environment variables.   This is a documentation problem outside of the scope of the issue as stated in #7 and #90

[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:71]: #90 states that we duplicate everything, not trying to optimize.   we were saving optimizations and deletions for later releases

File: internal/pkg/util/env/create.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: @cclerget "overridden" means "by prefixed environment variable".  is there another type of override that needs to be distinguished?   do we want a hyperlink to the user manual?   note that "prefixed environment variables" can come from `--env` `--env-file` command line options as well as apptainer config options and we do not have enough state preserved to indicate which source created the override.

[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: we also do not have clearly documented rules for order of precedence if the override is defined in multiple places.  ie: which wins when `--env key=val1,key=val2`, or key provided with val3 inside of `--env-file` ?

[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: this really seems out of scope for the current issue

[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: new issue could be `provide traceability for source of environment variable override`, as well as additional warnings when there are conflicting overrides

File: internal/pkg/util/env/create.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cclerget) [reply internal/pkg/util/env/create.go:71]: I'm focusing on renaming of `SING_USER_DEFINED_XXX` variables which could be renamed to whatever without impacting end users, not optimization, `SING_USER_DEFINED_XXX` variables are internal to runtime and are not propagated to container process as is, they are used to alter PATH environment variable (this is a legacy behavior), that's why you could just keep `setKeyIfNotAlreadyOverriden(g, envKeys, e[0], "APPTAINER_USER_DEFINED_XXX", e[1])` lines and remove lines with `SING_USER_DEFINED` and rename accordingly in the `RuntimeVars` embedded script

File: internal/pkg/util/env/create.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:132]: note that another source of override, depending on shell being used can be
```
SINGULARITYENV_KEY=foo apptainer ...
```

or
```
SINGULARITYENV_KEY=foo
apptainer ...
```

or
```
apptainer --env KEY=foo
```

all produce the same result.   The only tool I know of that keeps enough state to make the proper distinction is `make`

File: internal/pkg/util/env/create.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go]: removed the new variables.  out of scope for #7 and #90

File: internal/pkg/util/env/create.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:129]: per slack convo, settled on message `KEY already has defined value [%x], will not use value [%y] from parent environment`

[the remark this answers is not in the document]
Comment (PaulCharlton) [reply internal/pkg/util/env/create.go:129]: revised

**Review by cclerget — APPROVED** [review]

**Pull request merged**